src/CoreWCF.Primitives/src/CoreWCF/Channels/ByteStreamMessage.cs

```diff
@@ -0,0 +1,618 @@
+        {
+            return new InternalByteStreamMessage(httpResponseMessage, quotas, true); // moveBodyReaderToContent is true, for consistency with the other implementations of Message (including the Message base class itself)
+        }
+
+        internal static Message CreateMessage(ByteStreamBufferedMessageData bufferedMessageData, XmlDictionaryReaderQuotas quotas, bool moveBodyReaderToContent)
+        {
+            return new InternalByteStreamMessage(bufferedMessageData, quotas, moveBodyReaderToContent);
+        }
+
+        internal static bool IsInternalByteStreamMessage(Message message)
+        {
+            Fx.Assert(message != null, "message should not be null");
+            return message is InternalByteStreamMessage;
+        }
+
+        internal class InternalByteStreamMessage : Message
+        {
+            private BodyWriter _bodyWriter;
+            private readonly MessageHeaders _headers;
+            private readonly MessageProperties _properties;
+            private XmlByteStreamReader _reader;
+
+            /// <summary>
+            /// If set to true, OnGetReaderAtBodyContents() calls MoveToContent() on the reader before returning it.
+            /// If set to false, the reader is positioned on None, just before the root element of the body message.
+            /// </summary>
+            /// <remarks>
+            /// We use this flag to preserve compatibility between .net 4.0 (or previous) and .net 4.5 (or later).
+            ///
+            /// In .net 4.0:
+            /// - WebMessageEncodingBindingElement uses a raw encoder, different than ByteStreamMessageEncoder.
+            /// - ByteStreamMessageEncodingBindingElement uses the ByteStreamMessageEncoder.
+            /// - When the WebMessageEncodingBindingElement is used, the Message.GetReaderAtBodyContents() method returns 
+            ///   an XmlDictionaryReader positioned initially on content (the root element of the xml); that's because MoveToContent() is called
+            ///   on the reader before it's returned. 
+            /// - When the ByteStreamMessageEncodingBindingElement is used, the Message.GetReaderAtBodyContents() method returns an 
+            ///   XmlDictionaryReader positioned initially on None (just before the root element).
+            /// 
+            /// In .net 4.5:
+            /// - Both WebMessageEncodingBindingElement and ByteStreamMessageEncodingBindingElement use the ByteStreamMessageEncoder.
+            /// - So we need the ByteStreamMessageEncoder to call MoveToContent() when used by WebMessageEncodingBindingElement, and not do so
+            ///   when used by the ByteStreamMessageEncodingBindingElement.
+            /// - Preserving the compatibility with 4.0 is important especially because 4.5 is an in-place upgrade of 4.0.
+            /// 
+            /// See 252277 @ CSDMain for other info.
+            /// </remarks>
+            private readonly bool _moveBodyReaderToContent;
+
```

Comment:
This isn't needed any more. Can you make the code only do the newer behavior?

Reply:
This is addressed in b0c21a3804140f47b064720903835a1427a24357.